src/Xrefcheck/System.hs

```diff
@@ -41,50 +44,42 @@ askWithinCI = lookupEnv "CI" <&> \case
   Just (map C.toLower -> "true") -> True
   _ -> False
 
--- | Relative POSIX file path with some normalizations applied.
+-- | Relative file path with POSIX path separators.
 --
 -- This type exist in contrast to 'FilePath' which, in this project,
--- is used for platform-dependent filepaths and related filesystem
+-- is used for platform-dependent file paths and related filesystem
 -- IO operations.
 --
--- It should be created from a platform-dependent 'FilePath' via
+-- It should be created from a platform-dependent file path via
 -- 'mkRelPosixLink'.
 newtype RelPosixLink = UnsafeRelPosixLink
   { unRelPosixLink :: Text
   } deriving newtype (Show, Eq, Ord, NFData, Buildable, Pretty)
 
--- | Create a POSIX file path from a platform-dependent 'FilePath'.
---
--- Applies the following normalizations:
---
---  * Drop trailing path separator.
---
---  * FilePath.normalise, which is platform-dependent.
---
---  * Expand '.' and '..' indirections syntactically.
---
-mkRelPosixLink :: FilePath -> RelPosixLink
+-- | Create a POSIX file path from a platform-dependent one.
+mkRelPosixLink :: Text -> RelPosixLink
 mkRelPosixLink = UnsafeRelPosixLink
-  . fromString
   . withPathSeparator FPP.pathSeparator
```

Comment:
Btw, do we check somewhere that markdown links should not use backslash as path separator?

Reply:
We are not currently checking it but, just in case, here `\` is replaced by `/`. Should we check it and consider backslash as path separator usages invalid?

Reply:
I think yes, since GitHub and GitLab think so.

So we need two versions of `mkRelPosixLink` - one that accepts `FilePath` and does the replacement, and another one that accepts `Text` and fails on `\` encounters.

---

One other but related thing I wonder is - under Windows, will we get normal filepaths or they will be lower-cased? :joy:  Hope the tests will show us.

Reply:
I have changed the following:

- The `RelPosixLink` constructor is now exposed and not unsafe to use.
- `mkRelPosixLink` is now for building a `RelPosixLink ` from a `FilePath`.

After trying some approaches to this, I think that we should not forbid `\` characters in `GitHub` and `GitLab` links. Both of them seem to allow files named like `a\a.md`, and references to them work as expected. However, I tried to create a Golden test case for this and encountered some problems:

- When a file name contains some character like `\`, its corresponding line in the `git-ls files` command appears surrounded by double quotes and with that character escaped, and Xrefcheck considers that surrounding double quotes as being part of the file name, so it is invalid and cannot be read. Maybe we could use `String`'s `read` when the line starts with `"`.
- If, with that change, we managed to make Xrefcheck work with these file names, I think that we would have a problem in Windows:

File `a.md`
```
[Ref to a\a](a\a.md)
```

File `a\a.md`
```
[Ref to a.md](a.md)
```

This would be fine in GitHub, GitLab, Linux and macOS, but in Windows that last `a.md` reference would fail because it should be `../a.md` instead. The first one to `a\a.md` would work though.

Do you think that this deserves creating an issue?

Reply:
> After trying some approaches to this, I think that we should not forbid \ characters in GitHub and GitLab links. Both of them seem to allow files named like a\a.md, and references to them work as expected

Oh, that true. Allowing `\` is also an option.

Hmm, but still most of the `\` usages are likely mistakes introduced by Windows users. Perhaps to handle this, it would be worth extending the "file not found" error message in case it contains backshashes, say that likely a mistake takes place?

> When a file name contains some character like \, its corresponding line in the git ls-files command appears surrounded by double quotes

Interesting. I believe that this happens not only for `\` and so should generally be fixed. Could you do that in a separate commit? (and feel free to put that commit at the beginning of your PR)

AFAIS [from docs](https://git-scm.com/docs/git-ls-files#_output) we can also pass `-z` option, that would simplify parsing.

> If, with that change, we managed to make Xrefcheck work with these file names, I think that we would have a problem in Windows:

I wanted to say that Markdown's `a\a.md` should then correspond to Window's `a\\a.md` (or something like that) and vice versa, but then found out that `\` in filenames under Windows is not allowed, and including such file under Linux in git then can cause severe problems for Windows users :thinking:  I.e. `a\a.md` file will simplify get rejected by OS on `git clone`.

So maybe we don't even need to consider that case under Windows.

Reply:
Done!

> So maybe we don't even need to consider that case under Windows

It sounds good then, but it also means that I cannot add a test case with a `a\a.md` named file. Right?

> Perhaps to handle this, it would be worth extending the "file not found" error message

I think that it is a good idea and a simple thing to add 👍

Reply:
> It sounds good then, but it also means that I cannot add a test case with a a\a.md named file. Right?

Ugh, fair.

Mmm, this seems like a justified case to start using OS-dependent bats tests :thinking:  Like, set up some env variable in CI config, and rely on it in the test?